Signal an mxp event as an element is received

gcms opened the pull request:
- Enables to use scripts to handle elements received through MXP, such as NPCs' and objects' names and properties.

<!-- Keep the title short & concise so your mom can understand it -->
#### Brief overview of PR changes/additions

#### Motivation for adding to Mudlet

#### Other info (issues closed, discussion etc)

Commit:
Signal an mxp event as an element is received

- Enables to use scripts to handle elements received through MXP, such as NPCs' and objects' names andp properties.

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -14778,6 +14778,67 @@ void TLuaInterpreter::setAtcpTable(const QString& var, const QString& arg)
+        lua_setglobal(L, "mxp");
+        lua_getglobal(L, "mxp");
+        if (!lua_istable(L, -1)) {
+            qDebug() << "ERROR: mxp table not defined";
```
```suggestion
            qDebug() << "ERROR: mxp table could not be defined";
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -14778,6 +14778,67 @@ void TLuaInterpreter::setAtcpTable(const QString& var, const QString& arg)
+        lua_newtable(L);
+        lua_setfield(L, -2, type.toUtf8().toLower().constData());
+        if (!lua_istable(L, -1)) {
+            qDebug() << "ERROR: 'mxp." << type << "' table not defined";
```
```suggestion
            qDebug() << "ERROR: 'mxp." << type << "' table could not defined";
```

Kebap replied:
```suggestion
            qDebug() << "ERROR: 'mxp." << type << "' table could not be defined";
```

vadi2 reviewed (commented):
Very nice! Thanks for this. It works well!

![image](https://user-images.githubusercontent.com/110988/79187267-1ed9dd80-7e1c-11ea-95ef-3b877ae40969.png)

Could you make it be more idiomatic Lua, 1-indexed?

gcms commented:
> Very nice! Thanks for this. It works well!
> 
> ![image](https://user-images.githubusercontent.com/110988/79187267-1ed9dd80-7e1c-11ea-95ef-3b877ae40969.png)
> 
> Could you make it be more idiomatic Lua, 1-indexed?

Ok, I will do it. As for the error log I think we could leave it as it is, as it follows the pattern of the log messages for other protocols. Maybe it's better to keep them all consistent.

Commit:
Updated MXP event actions table to 1-based index

vadi2 reviewed (approved):
Cool thanks. Want to add support for mxp color while you're at it :wink:

gcms commented:
> Cool thanks. Want to add support for mxp color while you're at it

I just started using Mudlet yesterday and I'm still struggling on setting up the maps. Maybe if you give me some hints on where in the code I should look for and reference documentation on MXP I could give it a try.

vadi2 commented:
https://www.zuggsoft.com/zmud/mxp.htm that would be the <color> tag. Achaea uses it in it's QW output.

Join the discord and we'd be happy to help with the map!

Commit:
Fixed bug in MXP event lua stack management

- In some cases element attributes were written directly to the global mxp table instead of the event specific table.

vadi2 commented:
Could you also add documentation for this?

gcms commented:
> Could you also add documentation for this?

Do you mean code documentation or user documentation?

BTW I was looking at the code for implementing MXP color support, but I think it's useful to do some refactoring before, to extract the MXP handling behavior from the TBuffer. The TBuffer class is way too overloaded, so many responsibilities there, which makes it difficult to see the side effects.

I could do some reorganization, but I would like to know how can I do it in a way that avoids disrupting the development flow. If you don't mind I could do some baby steps refactoring and incrementally submit them as pull requests.

vadi2 commented:
User documentation, like we have it with GMCP.

Baby steps are definitely preferred, smaller PRs are easier to digest.

gcms commented:
> User documentation, like we have it with GMCP.
> 
> Baby steps are definitely preferred, smaller PRs are easier to digest.

Ok. I will start with something small to get some feedback from you. As for the documentation, is it in the code base?

vadi2 commented:
Sorry, I should have linked - https://wiki.mudlet.org/w/Manual:Technical_Manual#Supported_Protocols. It's up on the wiki as that has less friction to contribution for the non-coders.

vadi2 commented:
There's something odd going on - newlines printed inside `mxp.` events aren't interpreted as newlines. They're still interpreted fine when printed from other Mudlet-raised events, like `gmcp.`. Do you get the same?

![Selection_527](https://user-images.githubusercontent.com/110988/79629773-6c23bb00-814c-11ea-91d4-0bcf6a68597b.png)

```lua
function testmxpsend()
  print("Text for the link is: ".. mxp.send.href.."\n\n")
  print("Commands to do are: ")
  display(mxp.send.actions)
end
```

gcms committed:
Fix problems with new lines missing on MXP event

- The dity fix was to keep events in queue and print them from outside the TBuffer

vadi2 commented:
Hmm, the GMCP/MSDP don't require such a message queue - maybe we don't need to treat MXP specially here. I'll see if I can help you find it.

vadi2 commented:
So using this fixes it, but this isn't an issue players should have to concern themselves with, newlines should work out of the box:

```lua
function testmxpsend()
  print("Text for the link is: ".. mxp.send.href.."\n\n")
  print("Commands to do are: ")
  display(mxp.send.actions)
  wrapLine("main", getLineCount())
end

vadi2 commented:
Nope, your solutions looks good. Just two things - could you use a [qqueue](https://doc.qt.io/qt-5/qqueue.html) for this? and the formatting of tbuffer.h - did you use the `.clang-format` at the root of the repo? Using it I get a different result.

Thanks a lot :smiley:

gcms committed:
Replaced QList by QQueue in MXP event queue

vadi2 commented:
Cool, thanks a lot for this :)

vadi2 commented:
I've documented this for you here - https://wiki.mudlet.org/w/Manual:Supported_Protocols#MXP
